### PR TITLE
osmomath(log/CL): ln(x), log_1.0001(x), log_custom(x)

### PR DESCRIPTION
**Pull request by RusAkh**
Closes: #3132 

## What is the purpose of the change

`log` functions

## Testing and Verifying

tests added (2 test cases end with too big absolute error, I commented them in code)

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? no

### CODE REVIEW

File: osmomath/decimal_test.go [hunk 1]
```diff
@@ -739,3 +739,263 @@ func (s *decimalTestSuite) TestLog2() {
...
+		"log_0.2(0.99) = 0.006244624769837438271878639001855450": {
+			initialValue: MustNewDecFromStr("0.99"),
+			base:         MustNewDecFromStr("0.2"),
+			// From: https://www.wolframalpha.com/input?i=log_0.2%280.99%29+to+34+digits
+			expected: MustNewDecFromStr("0.006244624769837438271878639001855450"),
+		},
+		// THIS TEST CASE FAILS WITH 0.722861284485807159166346369373275313 ERROR
+		// "log_0.0001(500000) = -1.424742501084004701196565276318876743": {
+		// 	initialValue: NewDecWithPrec(1, 4),
+		// 	base:         NewBigDec(500000),
```
Comment (p0mvn) [inline osmomath/decimal_test.go:942]: Shouldn't `base` and `initialValue` be switched?

File: osmomath/decimal_test.go [hunk 2]
```diff
@@ -739,3 +739,263 @@ func (s *decimalTestSuite) TestLog2() {
...
+		// 	// From: https://www.wolframalpha.com/input?i=log_0.0001%28500000%29+to+37+digits
+		// 	expected: MustNewDecFromStr("-1.424742501084004701196565276318876743"),
+		// },
+
+		// THIS TEST CASE FAILS WITH 0.722861284485807159166346369373275311 ERROR
+		// "log_500000(0.0001) = -0.701881216598197542030218906945601429": {
+		// 	initialValue: NewBigDec(500000),
+		// 	base:         NewDecWithPrec(1, 4),
```
Comment (p0mvn) [inline osmomath/decimal_test.go:950]: I suspect the same might be happening here

File: osmomath/decimal_test.go [hunk 3]
```diff
@@ -739,3 +739,263 @@ func (s *decimalTestSuite) TestLog2() {
...
+		},
+		// Fails with 0.000000000000000000000000000000007686!!!!!
+		"log_1.123{1024.987654321} = 59.760484327223888489694630378785099461": {
+			initialValue: NewDecWithPrec(1024987654321, 9),
+			base:         NewDecWithPrec(1123, 3),
+			// From: https://www.wolframalpha.com/input?i=log_1.123%281024.987654321%29+to+38+digits
+			expected: MustNewDecFromStr("59.760484327223888489694630378785099461"),
+		},
+		// Fails with 0.000000000000000000000000000000077705!!!!!
+		"log_1.123{912648174127941279170121098210.92821920190204131121} = 594.689327867863079177915648832621538986": {
```
Comment (p0mvn) [inline osmomath/decimal_test.go:979]: I think this is fine. Let's do the following:
- add a test case field `expectedErrorTolerance`
- select the minimal `expectedErrorTolerance` value for each test case

File: osmomath/decimal_test.go [hunk 4]
```diff
@@ -739,3 +739,282 @@ func (s *decimalTestSuite) TestLog2() {
...
+			expectedPanic: true,
+		},
+		"log_1.0001{0}; invalid; panic": {
+			initialValue:  ZeroDec(),
+			expectedPanic: true,
+		},
+		"log_1.0001{0.001} = -69081.006609899112313305835611219486392199": {
+			initialValue: MustNewDecFromStr("0.001"),
+			// From: https://www.wolframalpha.com/input?i=log_1.0001%280.001%29+to+41+digits
+			expectedErrTolerance: MustNewDecFromStr("0.000000000000000000000000000143031879"),
```
Comment (czarcas7ic) [inline osmomath/decimal_test.go:847]: Do we think it makes more sense to have varying error tolerance levels for each test case or just a single overarching one for all the tests? The later seems more readable but I could be missing some important reason why this is not being done.

**Review by czarcas7ic — APPROVED** [review]
Just have a comment on the error tolerance, all else LGTM!

File: osmomath/decimal_test.go (hunk 4 above)
Comment (p0mvn) [reply osmomath/decimal_test.go:847]: While I agree that the latter is more readable, I think it's helpful to know the error tolerance for some subsets of inputs that we can infer from test cases.

I would prefer this knowledge over readability here

**Review by p0mvn — APPROVED** [review]
Great work

**Pull request merged**